WIP Add user resource

antham opened the pull request.

alexkappa commented on auth0/resource_auth0_user_test.go:
```diff
@@ -0,0 +1,74 @@
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(testAccUserCreateUser, os.Getenv("AUTH0_CLIENT_ID")),
```
Could also create a client for the purposes of this test. See client grants tests or resource server.

antham commented:
Something is wrong with this PR, it was working 2 weeks ago but I think you did some changes and updating a profile doesn't work anymore, so I need to fix it.

antham commented on auth0/resource_auth0_user.go:
```diff
@@ -0,0 +1,163 @@
+					return "auth0|" + val.(string)
+				},
+			},
+			"conn": {
```
No we can't it's a reserved keyword

antham commented on auth0/resource_auth0_user.go:
```diff
@@ -0,0 +1,163 @@
+		if err := d.Set(k, v); err != nil {
+			return err
+		}
+	}
```
Which one do you mean ?

alexkappa replied:
Any one of [`updateXXX`](https://github.com/yieldr/terraform-provider-auth0/blob/d0a2ce6d27a0fecb734f588ca945636288692981/auth0/resource_auth0_client.go#L274) and [`buildXXX`](https://github.com/yieldr/terraform-provider-auth0/blob/d0a2ce6d27a0fecb734f588ca945636288692981/auth0/resource_auth0_client.go#L289) for example.

antham replied:
I'm sorry but I still don't get the point, you mean why update, build and read don't follow conventions you defined ? If so which convention exactly ?

alexkappa replied:
Yes exactly, the convention I just linked. Comparing the two seems very different. All the resources follow this convention and I'd prefer to keep that.

antham replied:
It's different because you don't check the error coming from Set function. I can remove this but it's not a good thing not checking errors coming from a function, it could lead to unexpected troubles hard to debug.

antham commented on auth0/resource_auth0_user.go:
```diff
@@ -0,0 +1,163 @@
+		"email":          {Email: String(d, "email"), Connection: String(d, "conn")},
+		"email_verified": {EmailVerified: Bool(d, "email_verified"), Connection: String(d, "conn")},
+	} {
+		if d.HasChange(k) {
```
It was to get ride of this bunch of error :
<img width="454" alt="capture d ecran 2018-10-21 a 01 15 42" src="https://user-images.githubusercontent.com/4854264/47261403-027f0e00-d4cf-11e8-9fa5-0665c7c81eba.png">

alexkappa replied:
Then maybe a good application for [`ConflictsWith`](https://github.com/yieldr/terraform-provider-auth0/blob/d0a2ce6d27a0fecb734f588ca945636288692981/vendor/github.com/hashicorp/terraform/helper/schema/schema.go#L170)?

antham replied:
I don't really get how you use that and for which purpose exactly. But you still have to update partially if you want to have a correct overall behaviour.

alexkappa replied:
`ConflictsWith` would disallow specifying one if the other is specified. For example:

```
"email": {
  Type:     schema.TypeString,
  Optional: true,
  ConflictsWith: []string{"password"},
},
"password": {
  Type:     schema.TypeString,
  Optional: true,
  ConflictsWith: []string{"email"},
}
```

Would not allow you to set `password` if `email` is set.

To see if I get this right, is it allowed to specify `username` and `email` but not at the same time?

antham replied:
You can't change in the same time username, email, email_verified and password. I think I get the point but in my opinion, we don't want this behaviour, we want to be able to define everything in one go, because you want to be able for instance to define a username and an email in your terraform and having changes reflected in one apply without having to mess around with your config, you want everything to be managed in the background.

antham commented:
To solve my problem I changed the User struct in go-auth0, I switched AppMetadata and UserMetadata as pointer to map[string]interface. My guess is that it wasn't working because in the actual implementation when UserMetadata or AppMetadata are empty, they are not sent because of the omitempty clause, so you get a validation error because a request is sent but with not payload. But when we set UserMetadata or AppMetadata to be pointers, an empty pointer is nil otherwise, it will dump the empty map, see : https://play.golang.org/p/OTjntUIibkX. Is it ok if do a PR to change go-auth0 implementation regarding those fields ?

antham commented:
@alexkappa

antham commented:
@alexkappa did you see my ping ?

alexkappa commented:
Hi @antham, sorry I was travelling this past two weeks and havent had time to check. I’ll take a look and get back to you soon

antham commented:
Yep no problem.

antham commented:
I wasn't reluctant to do any changes in this PR, I was only expecting your answer to my last comments before changing anything @alexkappa

antham commented:
Ah ok sorry, my apologize as well. So what should I do ?

alexkappa commented:
I'm managed to get it to work reasonably well with the conditions auth0 applies during validation. Would you like to give some feedback on that pr? If we're happy with that, we can either merge there, or here. Your commits are there either way.
